src/metabase/legacy_mbql/normalize.cljc

```diff
@@ -910,6 +912,60 @@
                       {:query query}
                       e)))))
 
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                             LEGACY FILTER CLAUSES                                              |
+;;; +----------------------------------------------------------------------------------------------------------------+
+
+(defn- temporal-unit
+  [field]
+  (-> field mbql.u/field-options :temporal-unit))
+
+(defn- temporal-unit-is?
+  [field units]
+  (contains? units (temporal-unit field)))
```

Comment:
I think a slicker factoring of these `:guard`s would be:

`(field :guard (every-pred mbql.preds/Field? (temporal-unit-is? #{units...})))`

where that version of `temporal-unit-is?` returns a function:

```clojure
(defn- temporal-unit-is? [units]
  #(units (temporal-unit %)))
```
(Sets are functions and it's idiomatic to use them like that rather than call `contains?` explicitly.)